app/policies/work_package_policy.rb

```diff
@@ -94,4 +95,12 @@ def comment_allowed?(work_package)
   def assign_version_allowed?(work_package)
     user.allowed_in_project?(:assign_versions, work_package.project)
   end
+
+  def change_status_allowed?(work_package)
+    @change_status_cache ||= Hash.new do |hash, project|
+      hash[project] = user.allowed_in_project?(:change_work_package_status, work_package.project)
+    end
+
+    @change_status_cache[work_package.project]
+  end
```

Comment:
As user Alice, I tried sharing a work package with edit rights to user Bob having no access to a project.

As expected Bob now have limted access to the project, and can only see and edit the work package that Alice shared. Bob can edit everything but the status. Yet, the `WorkPackageRole` with name `Work Package Editor` has the permission `:change_work_package_status` being listed. (You can run `WorkPackageRole.find_by(builtin: Role::BUILTIN_WORK_PACKAGE_EDITOR).permissions` in a console to check).

I think it's lacking a work package level permission check, checking that either `user.allowed_in_project?(:change_work_package_status, work_package.project)` or `user.allowed_in_work_package?(:change_work_package_status, work_package)` returns true.

I tried changing the implementation to a naive `user.allowed_in_project?(:change_work_package_status, work_package.project) || user.allowed_in_work_package?(:change_work_package_status, work_package)` but it did not work. I suppose the schema is also interfering as it does its checks at the project level.

I am not comfortable yet with the new work package permissions system. I'll ask for some help here.

Reply:
I tried my best, but even while implementing this I was still learning more and more about OP's secrets, and I'm not aware of all the features obviously. So I'll welcome some more direct help with this, if possible.

Reply:
I can relate to that, and the work package permissions is very new. No problem, we'll see how it can be fixed.
And if `:change_work_package_status` is not more distinct from `:edit_work_package`, that would solve the issue as well.

Reply:
It has been decided that `:edit_work_package` will encompass `:change_work_package_status`, so this problem will go away.

Just for information, if it was not, then something like `user.allowed_in_any_work_package?(:change_work_package_status)` would have been needed for the schema, because the schema and the writability of each field is scoped to the project. It would allow the user to click any status, and upon clicking the update_form endpoint would be queried, and that would return the effective permissions for the work package, forbidding or allowing the user to modify the field.

That's not really nice but that's how it should have been implemented.